### ISSUE TITLE
[no-unused-vars] feature request: an option to ignore inferred type variables in conditionals

### ISSUE THREAD

**Issue by susisu** [typescript-eslint/typescript-eslint#2584]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
export type Fns<T, U> = {
  f: (x: T) => T;
  g: (x: U) => U;
};
export type FParam<X> = X extends Fns<infer T, infer U> ? T : never;
```

**Expected Result**

No errors reported, because although the inferred type variable `U` is unused, it is unavoidable.
Using [varsIgnorePattern](https://eslint.org/docs/rules/no-unused-vars#varsignorepattern) can suppress the error, but it also suppresses other errors which might be unexpected.
I think an option like [ignoreRestSiblings](https://eslint.org/docs/rules/no-unused-vars#ignorerestsiblings), or something like `typeVarsIgnorePattern` would be better.

**Actual Result**

```
  5:54  error  'U' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.1.1` |
| `@typescript-eslint/parser`        | `4.1.1` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.9.0` |
| `node`                             | `14.6.0` |

**Comment by bradzacher** [comment]
> it is unavoidable

ways to avoid it:

```ts
export type FParam<X> = X extends Fns<infer T, unknown> ? T : never;
export type FParam<X> = X extends Fns<infer T, never> ? T : never;
export type FParam<X> = X extends Fns<infer T, any> ? T : never;
```

The inferred type is unused and it is unnecessary.
We will not add an option for this case.

**Comment by susisu** [comment]
@bradzacher Hi, thank you for your response!
Yes, `any` works, but it is warned by `no-explicit-any` (which is a recommended rule) so I want to avoid using it (or I have to write `eslint-ignore` after all).
Neither `unknown` nor `never` do not work in my case, because the parameter `U` is in invariant position.

**Comment by bradzacher** [comment]
Three solutions:

1) Use `any`
`any` should be avoided in many places.
This is a safe use and is literally what it's designed for.
Using `any` here will not introduce any unsafety in your codebase because you literally don't care about the value of the second generic parameter.

2) Use `varsIgnorePattern` set to something like `^_` (most codebases have it setup this way) and rename the parameter to `infer _U`.

3) Use an `eslint-disable` comment.

We will not be adding an option for this use case.

**Comment by susisu** [comment]
Ok, I'll use `varsIgnorePattern` for this case for now.
But because type variables are somewhat different from value-level variables, I would be happy if typescript-eslint have distinct options for them (as discussed in #2337 for type parameters and function arguments).
Anyway, thank you for your answer.
